File: routes/objc-to-swift/certification/artifacts/module/source-runtime/000/source_harness.m
#import <Foundation/Foundation.h>
#include <math.h>
#include <stdint.h>
#include <string.h>
#import "equivalence_module.m"

static __attribute__((unused)) uint64_t ElmosHarnessFP64Bits(double value) {
    uint64_t bits = 0;
    memcpy(&bits, &value, sizeof(bits));
    return bits;
}

static __attribute__((unused)) BOOL ElmosHarnessSameFP64(double left, double right) {
    return (isnan(left) && isnan(right)) ||
           ElmosHarnessFP64Bits(left) == ElmosHarnessFP64Bits(right);
}

static __attribute__((unused)) NSString *ElmosHarnessFP64(double value) {
    return [NSString stringWithFormat:@"%016llx", (unsigned long long)ElmosHarnessFP64Bits(value)];
}

static __attribute__((unused)) NSString *ElmosHarnessHexUTF8(NSString *value) {
    NSData *data = [value dataUsingEncoding:NSUTF8StringEncoding];
    const unsigned char *bytes = data.bytes;
    NSMutableString *result = [NSMutableString stringWithCapacity:data.length * 2];
    for (NSUInteger index = 0; index < data.length; index++) {
        [result appendFormat:@"%02x", (unsigned int)bytes[index]];
    }
    return result;
}

int main() {
    @autoreleasepool {
        BOOL actual_0 = both(YES, YES);
        BOOL expected_0 = YES;
        if (actual_0 != expected_0) return 1;
        printf("ELMOS_OBSERVATION\t0\tbool\t%s\n", actual_0 ? "true" : "false");
        BOOL actual_1 = both(YES, NO);
        BOOL expected_1 = NO;
        if (actual_1 != expected_1) return 2;
        printf("ELMOS_OBSERVATION\t1\tbool\t%s\n", actual_1 ? "true" : "false");
        BOOL actual_2 = both(NO, YES);
        BOOL expected_2 = NO;
        if (actual_2 != expected_2) return 3;
        printf("ELMOS_OBSERVATION\t2\tbool\t%s\n", actual_2 ? "true" : "false");
        BOOL actual_3 = both(NO, NO);
        BOOL expected_3 = NO;
        if (actual_3 != expected_3) return 4;
        printf("ELMOS_OBSERVATION\t3\tbool\t%s\n", actual_3 ? "true" : "false");
    }
    return 0;
}

File: routes/cpp-to-objc/certification/artifacts/module/source-runtime/003/source_harness.cpp
#include <cmath>
#include <cstdint>
#include <cstring>
#include <iomanip>
#include <iostream>
#include <limits>
#include <sstream>
#include <string>
#include "equivalence_module.cpp"

[[maybe_unused]] static std::uint64_t elmos_harness_fp64_bits(double value) {
    std::uint64_t bits = 0;
    static_assert(sizeof(bits) == sizeof(value));
    std::memcpy(&bits, &value, sizeof(bits));
    return bits;
}

[[maybe_unused]] static bool elmos_harness_same_fp64(double left, double right) {
    return (std::isnan(left) && std::isnan(right)) ||
           elmos_harness_fp64_bits(left) == elmos_harness_fp64_bits(right);
}

[[maybe_unused]] static std::string elmos_harness_fp64(double value) {
    std::ostringstream stream;
    stream << std::hex << std::setfill('0') << std::setw(16)
           << elmos_harness_fp64_bits(value);
    return stream.str();
}

[[maybe_unused]] static std::string elmos_harness_hex_utf8(const std::string &value) {
    static constexpr char digits[] = "0123456789abcdef";
    std::string encoded;
    encoded.reserve(value.size() * 2);
    for (const unsigned char byte : value) {
        encoded.push_back(digits[byte >> 4]);
        encoded.push_back(digits[byte & 0x0f]);
    }
    return encoded;
}

int main() {
    const auto actual_0 = clampNumber(-10.5, 0.0, 100.0);
    const auto expected_0 = 0.0;
    if (!elmos_harness_same_fp64(actual_0, expected_0)) return 1;
    std::cout << "ELMOS_OBSERVATION\t0\tfp64-hex\t" << elmos_harness_fp64(actual_0) << "\n";
    const auto actual_1 = clampNumber(55.25, 0.0, 100.0);
    const auto expected_1 = 55.25;
    if (!elmos_harness_same_fp64(actual_1, expected_1)) return 2;
    std::cout << "ELMOS_OBSERVATION\t1\tfp64-hex\t" << elmos_harness_fp64(actual_1) << "\n";
    const auto actual_2 = clampNumber(101.5, 0.0, 100.0);
    const auto expected_2 = 100.0;
    if (!elmos_harness_same_fp64(actual_2, expected_2)) return 3;
    std::cout << "ELMOS_OBSERVATION\t2\tfp64-hex\t" << elmos_harness_fp64(actual_2) << "\n";
    const auto actual_3 = clampNumber(-0.0, -1.0, 1.0);
    const auto expected_3 = -0.0;
    if (!elmos_harness_same_fp64(actual_3, expected_3)) return 4;
    std::cout << "ELMOS_OBSERVATION\t3\tfp64-hex\t" << elmos_harness_fp64(actual_3) << "\n";
    const auto actual_4 = clampNumber(0.0, -0.0, 1.0);
    const auto expected_4 = 0.0;
    if (!elmos_harness_same_fp64(actual_4, expected_4)) return 5;
    std::cout << "ELMOS_OBSERVATION\t4\tfp64-hex\t" << elmos_harness_fp64(actual_4) << "\n";
    return 0;
}
